tests/lib/k8s/caasp.py

```diff
@@ -0,0 +1,135 @@
+        self.clusterpath = os.path.join(hardware.working_dir, 'cluster')
+        self._kubeconfig = os.path.join(self.hardware.working_dir,
+                                        'admin.conf')
+        self.ssh_agent()
+
+    def ssh_agent(self):
+        try:
+            res = subprocess.run(['ssh-agent'], check=True,
+                                 capture_output=True)
+        except subprocess.CalledProcessError:
+            msg = 'Failed to start ssh agent'
+            logger.error(msg)
+            raise Exception(msg)
+
+        self.SSH_AUTH_SOCK = res.stdout.decode(
+            'utf-8').split(';')[0].split('=')[1]
+        self.SSH_AGENT_PID = res.stdout.decode(
+            'utf-8').split(';')[2].split('=')[1]
+        os.environ['SSH_AUTH_SOCK'] = self.SSH_AUTH_SOCK
+        os.environ['SSH_AGENT_PID'] = self.SSH_AGENT_PID
+        try:
+            res = subprocess.run(['ssh-add', self.hardware._private_key],
+                                 check=True)
+        except subprocess.CalledProcessError:
+            msg = 'Failed to add keys to agent'
+            logger.error(msg)
+            raise Exception(msg)
+
+    def destroy(self, skip=False):
+        logger.info(f"kube destroy on hardware {self.hardware}")
+        if skip:
+            # We can skip in most cases since the nodes themselves will be
+            # destroyed instead.
+            return
+        # This kills the SSH_AGENT_PID agent
+        try:
+            subprocess.run(['ssh-agent', '-k'], check=True)
+        except subprocess.CalledProcessError:
+            logger.exception(f'Killing ssh-agent with PID \
+{self.SSH_AGENT_PID} failed')
+
+        # TODO(jhesketh): Uninstall kubernetes
+
+    def install_kubernetes(self):
+        super().install_kubernetes()
+        self.hardware.execute_ansible_play_raw('playbook_caasp.yaml')
+        self.hardware.get_masters()
+        self.hardware.get_workers()
+        self._caasp_init()
+        with cd(self.clusterpath):
+            self._caasp_bootstrap()
+        with cd(self.clusterpath):
+            self._caasp_join()
+
+    def _caasp_init(self):
+        try:
+            res = subprocess.run(
+                ['skuba', 'cluster', 'init', '--control-plane',
+                 self.hardware.masters[0].get_ssh_ip(),
+                 self.clusterpath], check=True)
+            logger.debug(res.args)
+        except subprocess.CalledProcessError:
+            msg = 'Cluster init step failed'
+            logger.error(msg)
+            raise Exception(msg)
+
+    def _caasp_bootstrap(self):
+        try:
+            res = subprocess.run(
+                ['skuba', 'node', 'bootstrap', '--user', 'sles', '--sudo',
+                 '--target', self.hardware.masters[0].get_ssh_ip(),
+                 self.hardware.masters[0].dnsname], check=True)
+            logger.debug(res.args)
+        except subprocess.CalledProcessError:
+            msg = 'Cluster bootsrap step failed'
```

Comment:
Having the full command in the log would be nice.

Reply:
Handled by logger.exception